cnxarchive/views/xpath.py

```diff
@@ -0,0 +1,147 @@
+import json
+
+import psycopg2
+import psycopg2.extras
+from pyramid import httpexceptions
+from lxml import etree
+from pyramid.threadlocal import get_current_registry, get_current_request
+from pyramid.view import view_config
+from .content import _get_content_json, HTML_WRAPPER
+
+from .. import config
+from ..database import (
+    SQL)
+from ..utils import (
+    IdentHashShortId, IdentHashMissingVersion,
+    split_ident_hash, COLLECTION_MIMETYPE
+    )
+from .helpers import get_uuid, get_latest_version
+
+
+@view_config(route_name='xpath-book', request_method='GET')
+def xpath_book(request):
+    """
+    Given a book UUID (and optional version) and escaped xpath string:
+
+    if Accept: 'application/xhtml+xml' is in the headers, returns the HTML tree of the book, with each module title a
+    link which returns the xpath query result for that page UUID (and version if supplied, else uses most recent), or
+
+    otherwise, returns a JSON object of results, each result containing:
+    module_name,
+    module_uuid,
+    xpath_results, an array of strings, each an individual xpath result.
+    """
+
+    if 'application/xhtml+xml' in request.headers.get('ACCEPT', ''):
+        return xpath_book_html(request)
+    else:
+        return execute_xpath(request, 'xpath')
+
+
+@view_config(route_name='xpath-page', request_method='GET', renderer='json')
+def xpath_page(request):
+    """Given a page UUID (and optional version), returns a JSON object of results, as in xpath_book()"""
+    return execute_xpath(request, 'xpath-module')
```

Comment:
Can this also generate an HTML list of links (like the book results)? (see `#${nearest_element_id}` comment below)